src/mon/OSDMonitor.cc

```diff
@@ -4385,7 +4385,8 @@ bool OSDMonitor::prepare_beacon(MonOpRequestRef op)
   osd_epochs[from] = beacon->version;
 
   for (const auto& pg : beacon->pgs) {
-    last_epoch_clean.report(pg, beacon->min_last_epoch_clean);
+    auto pg_num = osdmap.get_pg_pool(pg.pool())->get_pg_num();
```

Comment:
```suggestion
    if (auto* pool = osdmap.get_pg_pool(pg.pool()); pool != nullptr) {
      unsigned pg_num = pool->get_pg_num();
      // ...
    }
```

there is chance that the PG being reported has been removed by monitor, so `osdmap.get_pg_pool(pg.pool())` evaluates to `nullptr`, and hence causes segfault. see https://pulpito.ceph.com/kchai-2021-07-06_12:30:55-rados-wip-kefu-testing-2021-07-06-1559-1-distro-basic-smithi/, in which lots of failures were caused by the segfault in `OSDMonitor::prepare_update()`, when monitor was handling a beacon containing removed PG(s) from osd.

Reply:
the test result is much better with the proposed change. but the monitor still crash in couple tests. see https://pulpito.ceph.com/kchai-2021-07-07_05:44:54-rados-wip-kefu-testing-2021-07-07-1141-distro-basic-smithi/

```
   -15> 2021-07-07T06:17:44.986+0000 7ff899261700 10 mon.a@0(leader) e1  caps are allow *
   -14> 2021-07-07T06:17:44.986+0000 7ff899261700 10 mon.a@0(leader) e1  entity name 'osd.1' type 4
   -13> 2021-07-07T06:17:44.986+0000 7ff899261700 10 mon.a@0(leader) e1  mesg 0x55d9bda9d680 from v2:172.21.15.77:3301/0
   -12> 2021-07-07T06:17:44.986+0000 7ff899261700 20 mon.a@0(leader) e1 _ms_dispatch existing session 0x55d9be06f680 for osd.1
   -11> 2021-07-07T06:17:44.986+0000 7ff899261700 20 mon.a@0(leader) e1  entity_name osd.1 global_id 0 (none) caps allow *
   -10> 2021-07-07T06:17:44.986+0000 7ff899261700 10 mon.a@0(leader).paxosservice(osdmap 1..602) dispatch 0x55d9bda9d680 osd_beacon(pgs [2.8,3.0,2.0,3.9,3.39,3.31,3.19,3.1a,1.b,1.3,1.13,2.c,2.4,3.4,3.25,2
.e,2.6] lec 490 last_purged_snaps_scrub 2021-07-07T06:17:44.987351+0000 osd_beacon_report_interval 300 v602) v3 from osd.1 v2:172.21.15.77:6800/33521 con 0x55d9c03ce540
    -9> 2021-07-07T06:17:44.986+0000 7ff899261700  5 mon.a@0(leader).paxos(paxos active c 1005..1542) is_readable = 1 - now=2021-07-07T06:17:44.987784+0000 lease_expire=2021-07-07T06:17:49.866037+0000 has
 v0 lc 1542
    -8> 2021-07-07T06:17:44.986+0000 7ff899261700 10 mon.a@0(leader).osd e602 preprocess_query osd_beacon(pgs [2.8,3.0,2.0,3.9,3.39,3.31,3.19,3.1a,1.b,1.3,1.13,2.c,2.4,3.4,3.25,2.e,2.6] lec 490 last_purge
d_snaps_scrub 2021-07-07T06:17:44.987351+0000 osd_beacon_report_interval 300 v602) v3 from osd.1 v2:172.21.15.77:6800/33521
    -7> 2021-07-07T06:17:44.986+0000 7ff899261700 10 mon.a@0(leader) e1 no_reply to osd.1 v2:172.21.15.77:6800/33521 via v2:172.21.15.77:3301/0 for request osd_beacon(pgs [2.8,3.0,2.0,3.9,3.39,3.31,3.19,3
.1a,1.b,1.3,1.13,2.c,2.4,3.4,3.25,2.e,2.6] lec 490 last_purged_snaps_scrub 2021-07-07T06:17:44.987351+0000 osd_beacon_report_interval 300 v602) v3
    -6> 2021-07-07T06:17:44.986+0000 7ff899261700  1 -- [v2:172.21.15.77:3300/0,v1:172.21.15.77:6789/0] --> [v2:172.21.15.77:3301/0,v1:172.21.15.77:6790/0] -- route(no-reply tid 2297) v3 -- 0x55d9bd21c2c0
 con 0x55d9bcff0000
    -5> 2021-07-07T06:17:44.986+0000 7ff899261700 20 is_capable service=osd command= exec addr v2:172.21.15.77:6800/33521 on cap allow *
    -4> 2021-07-07T06:17:44.986+0000 7ff899261700 20  allow so far , doing grant allow *
    -3> 2021-07-07T06:17:44.986+0000 7ff899261700 20  allow all
    -2> 2021-07-07T06:17:44.986+0000 7ff899261700  7 mon.a@0(leader).osd e602 prepare_update osd_beacon(pgs [2.8,3.0,2.0,3.9,3.39,3.31,3.19,3.1a,1.b,1.3,1.13,2.c,2.4,3.4,3.25,2.e,2.6] lec 490 last_purged_
snaps_scrub 2021-07-07T06:17:44.987351+0000 osd_beacon_report_interval 300 v602) v3 from osd.1 v2:172.21.15.77:6800/33521
    -1> 2021-07-07T06:17:44.986+0000 7ff899261700 10 mon.a@0(leader).osd e602 prepare_beacon osd_beacon(pgs [2.8,3.0,2.0,3.9,3.39,3.31,3.19,3.1a,1.b,1.3,1.13,2.c,2.4,3.4,3.25,2.e,2.6] lec 490 last_purged_
snaps_scrub 2021-07-07T06:17:44.987351+0000 osd_beacon_report_interval 300 v602) v3 from osd.1
     0> 2021-07-07T06:17:44.988+0000 7ff899261700 -1 *** Caught signal (Aborted) **
 in thread 7ff899261700 thread_name:ms_dispatch
```

Reply:
Maybe the OSDs continue to send beacons for the deleted PGs (ps>=pg_num) while they are still being physically removed from the OSD?

Reply:
Or perhaps we need:
```
 void LastEpochClean::report(unsigned pg_num, const pg_t& pg,
                            epoch_t last_epoch_clean)
 {
-  auto& lec = report_by_pool[pg.pool()];
-  return lec.report(pg_num, pg.ps(), last_epoch_clean);
+  if (auto* pool = pg.pool(); pool != nullptr) {
+    auto& lec = report_by_pool[pg.pool()];
+    return lec.report(pg_num, pg.ps(), last_epoch_clean);
+  }
 }
 ```
 I will send a test now

Reply:
> Or perhaps we need:
> 
> ```
>  void LastEpochClean::report(unsigned pg_num, const pg_t& pg,
>                             epoch_t last_epoch_clean)
>  {
> -  auto& lec = report_by_pool[pg.pool()];
> -  return lec.report(pg_num, pg.ps(), last_epoch_clean);
> +  if (auto* pool = pg.pool(); pool != nullptr) {
> +    auto& lec = report_by_pool[pool];
> +    return lec.report(pg_num, pg.ps(), last_epoch_clean);
> +  }
>  }
> ```
> 
> I will send a test chnage now

ignore this.. totally wrong.
It seems conceivable that report_by_pool[pg.pool()] doesn't exist, because it could be remove_pool'd while we update a beacon, but this would not have regressed in this PR.

Reply:
> Maybe the OSDs continue to send beacons for the deleted PGs (ps>=pg_num) while they are still being physically removed from the OSD?

yeah, i think this is a promising theory to explain the crash. we autoscale the pg nowadays. so OSDs might be reporting removed PGs which belongs to an *existing* pool.